### PR TITLE
Add Cholesky factorization

### LINKED ISSUE #1164: Numerical Cholesky factorization
The existing LU factorization kernel can be modified to build a Cholesky factorization.

### PR DESCRIPTION
**Pull request by upsj**
This PR adds Cholesky factorization kernels and interface. It uses the elimination tree to wait for all dependencies to be fulfilled, before factorizing the row in the lower triangle and writing the resulting column into the (transposed) upper triangle

Closes #1164

### CODE REVIEW

**Comment by upsj** [comment]
format!

**Comment by upsj** [comment]
The performance regression compared to LU makes this currently unsuitable for GPUs, so I'll have to work on it a bit more

**Comment by upsj** [comment]
format-rebase!

File: dpcpp/factorization/cholesky_kernels.dp.cpp [hunk 1]
```diff
@@ -175,6 +174,49 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
...
+
+template <typename ValueType, typename IndexType>
+void factorize_ldl(std::shared_ptr<const DefaultExecutor> exec,
+                   const IndexType* lookup_offsets, const int64* lookup_descs,
+                   const int32* lookup_storage, const IndexType* diag_idxs,
+                   const IndexType* transpose_idxs,
+                   matrix::Csr<ValueType, IndexType>* factors,
+                   array<int>& tmp_storage) GKO_NOT_IMPLEMENTED;
+
+
```
Comment (upsj) [inline dpcpp/factorization/cholesky_kernels.dp.cpp:219]:
```suggestion
```

**Comment by upsj** [comment]
format!

**Review by fritzgoebel — APPROVED** [review]
LGTM. The test matrices appear to be getting a bit out of hand size wise.. Maybe we can in a separate effort try to have the test matrices stored in binary?

File: common/cuda_hip/factorization/cholesky_kernels.hpp.inc [hunk 2]
```diff
@@ -156,3 +144,272 @@ __launch_bounds__(default_block_size) void cholesky_symbolic_factorize_kernel(
...
+    __shared__ int ready[default_block_size];
+    ready[tidx] = 0;
+    __syncthreads();
+    if (tidx >= num_rows) {
+        return;
+    }
+
+    const auto child_begin = child_ptrs[num_rows];
+    const auto child_end = child_ptrs[num_rows + 1];
+    for (auto i = child_begin; i < child_end; i++) {
```
Comment (vasilisge0) [inline common/cuda_hip/factorization/cholesky_kernels.hpp.inc:301]: What is the purpose of this loop? Is something missing?

**Review by vasilisge0 — APPROVED** [review]
LGTM!

File: common/cuda_hip/factorization/cholesky_kernels.hpp.inc (hunk 2 above)
Comment (upsj) [reply common/cuda_hip/factorization/cholesky_kernels.hpp.inc:301]: oh right, I left some WIP code for postordering the tree on the GPU. Let me remove that, I can always recover it from the history

File: test/factorization/cholesky_kernels.cpp [hunk 3]
```diff
@@ -180,21 +214,303 @@ TYPED_TEST(Cholesky, KernelSymbolicFactorize)
...
+        gko::factorization::compute_elim_forest(mtx_chol.get(), forest);
+        gko::factorization::compute_elim_forest(dmtx_chol.get(), dforest);
+    }
+
+    void forall_matrices(std::function<void()> fn)
+    {
+        {
+            SCOPED_TRACE("ani1");
+            this->initialize_data(gko::matrices::location_ani1_mtx,
+                                  gko::matrices::location_ani1_chol_mtx);
```
[the remark this answers is not in the document]
Comment (upsj) [reply test/factorization/cholesky_kernels.cpp:347]: the Cholesky factorization of the lower triangle of `ani1`, i.e. `chol(A, 'lower')`

File: core/factorization/cholesky.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<LinOp> Cholesky<ValueType, IndexType>::generate_impl(
+    std::shared_ptr<const LinOp> system_matrix) const
+{
+    GKO_ASSERT_IS_SQUARE_MATRIX(system_matrix);
+    const auto exec = this->get_executor();
+    // TODO deal with non Csr matrices
+    const auto mtx = as<matrix_type>(system_matrix);
```
[the remark this answers is not in the document]
Comment (upsj) [reply core/factorization/cholesky.cpp:93]: right, thank you! I kind of dislike the design of the function, but it should be appropriate here.

**Pull request merged**